Add GPU sensor to Glances

fhoekstra opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
In [python-glances-api release 0.5.0](https://github.com/home-assistant-ecosystem/python-glances-api/releases/tag/0.5.0), Glances' GPU sensors are exposed so that Home Assistant can display and track them.
This change is about adding these GPU sensors to Home Assistant.
This fixes issue #74444


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #74444
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/30481

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

fhoekstra committed:
feat: add GPU sensor to Glances

fhoekstra commented on tests/components/glances/test_sensor.py:
```diff
@@ -43,6 +43,16 @@ async def test_sensor_states(hass: HomeAssistant) -> None:
+    if state := hass.states.get(sensor_name_gpu + "_temperature"):
+        assert state.state == str(test_data_gpu["temperature"])
+    if state := hass.states.get(sensor_name_gpu + "_fan_speed"):
+        assert state.state == str(test_data_gpu["fan_speed"])
```
@engrbm87 I am a bit surprised that the GPU memory is registered as a string in Home Assistant state, while the sensor for RAM memory use looks very similar and does not need the string cast in the test here. Any idea what's causing this?

engrbm87 replied:
I can see that the GPU sensors are mostly percentages. For the memory sensors there is actual memory and memory usage as percentage.
[This link](https://github.com/nicolargo/glances/blob/develop/glances/plugins/gpu/__init__.py) shows the UOM for the GPU sensors, please double check that they are the same in our code here.
As for the string issue I am not sure, what error are you getting in the test if you don't cast as str.

fhoekstra replied:
The HA `state` for the GPU sensors is a string, even though the input data (`HA_SENSOR_DATA` in glances_api) is int or float. Not sure why the sensor state becomes string, as the sensors above in this test seem to have the same sensor implementation but do have a *numerical* HA state (docker memory use for example).

engrbm87 replied:
I did some digging and found that the sensors tests added in PR #93542 is not using the correct `entity_id`. Because it is using an if statement to get the state variable the assertion is not getting executed which is why we are not getting an assertion error. I think the tests should be fixed in a preliminary PR.
As for the `str` casting, I think this is required because the entity state will always be a string.

fhoekstra commented:
I haven't been able to test locally so far as I am developing in VSCode DevContainer and haven't been able to expose my GPU to the DevContainer yet.

fhoekstra commented:
@engrbm87 This is almost done. I have completed local testing but am struggling to add the GPU to the new style of snapshot testing. Could you point me in the right direction for that?
Manually editing the test_sensor.ambr file does not seem to work for me  so far with this:
```
# name: test_sensor_states[sensor.0_0_0_0_gpu_0_nvidia_geforce_rtx_3080_temperature-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'device_class': 'temperature',
      'friendly_name': '0.0.0.0 GPU_0__NVIDIA GeForce RTX 3080 temperature',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': <UnitOfTemperature.CELSIUS: '°C'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.0_0_0_0_gpu_0_nvidia_geforce_rtx_3080_temperature',
    'last_changed': <ANY>,
    'last_updated': <ANY>,
    'state': 'unavailable',
  })
# ---
```

joostlek commented:
@fhoekstra, run the tests with `pytest ./tests/components/glances --snapshot-update` it will generate the snapshots

fhoekstra commented:
Thanks @joostlek , just did that. Wasn't that hard to find that syrupy is the library we use, would a PR for the docs with mentioning syrupy and snapshots be appreciated?

I added the e-mail above to my Github account according to the instructions of the cla-bot and rebased the branch, but the bot still does not seem happy.

joostlek commented:
https://developers.home-assistant.io/docs/development_testing#snapshot-testing :)

cohenchris commented:
What's the status on this? Seems it all checks out, just need somebody with write access to approve.

joostlek reviewed (changes requested):
Please rebase your branch. Translations have been added to Glances so this PR should add them too.

fhoekstra committed:
Add translations to Glances GPU sensor

fhoekstra commented:
@joostlek You're right, I dropped a commit somewhere. I did generate them now, but they look off: a few sensors don't have the GPU part of the entity ID it seems. Converting this back to draft until I can fix this

joostlek commented:
Did you generate translations beforehand?

joostlek commented:
`python -m script.translations develop --all`

fhoekstra commented:
I did, but nothing happened. Maybe I messed up the `translation_key`s (I copied them from existing sensors as much as possible)

fhoekstra committed:
Fix translations of GPU processor and memory usage

joostlek commented:
Something still feels off `GPU_0__NVIDIA GeForce RTX 3080 processor usage` doesnt look like a good name

fhoekstra commented:
I think it's all good now, added some new custom translation keys for processor and memory usage with the `{sensor_label}` template so the GPU name is inserted there and the entities can be distinguished based on that.

fhoekstra commented:
Shall I replace the (double) underscores with spaces?

joostlek commented:
Oh its getting the ID from the json. I don't really like the name like this (as we have the user friendly name, so I think its strange to use some GPU__0_). The only case where this would make sense if someone has 2 the same GPUs. I'm probably leaving this discussion for @engrbm87

fhoekstra commented on tests/components/glances/__init__.py:
```diff
@@ -207,4 +207,13 @@
         },
     },
+    "gpu": {
+        "GPU_0__NVIDIA GeForce RTX 3080": {
+            "name": "NVIDIA GeForce RTX 3080",
+            "temperature": 51,
+            "mem": 8.41064453125,
+            "proc": 26,
+            "fan_speed": 0,
+        }
+    },
```
this is what I get from the glancesapi library. Shall I put in some string manipulation to clean that up or request changes upstream?

fhoekstra commented:
In f041750 , I put in a proposal for removing the `GPU_0__` prefix if only a single GPU is present on a host. @engrbm87 , let me know what you think of it.

fhoekstra commented:
How about this:
* If there is only 1 GPU present on a host, remove the `GPU_0__` prefix
`GPU_0__Nvidia RTX 3080` -> `Nvidia RTX 3080`
* If there are multiple GPUs present on a host, turn the prefix into a suffix: 
`GPU_0__Nvidia RTX 3080` -> `Nvidia RTX 3080 (GPU 0)`
`GPU_1__Nvidia RTX 3070Ti` -> `Nvidia RTX 3070Ti (GPU 1)`

Advantages:
* no weird numbers when a host has only a single GPU, `friendly_name` is friendly
* entity IDs always start with the make and model of the GPU
* GPU index is present in entity ID if there are multiple GPUs
* Multiple identical GPUs can be told apart

Disadvantages:
* A GPU's entity ID changes if another GPU is added to a host
* ?

cohenchris commented:
> How about this?
> 
> * If there is only 1 GPU present on a host, remove the `GPU_0__` prefix
>   `GPU_0__Nvidia RTX 3080` -> `Nvidia RTX 3080`
> * If there are multiple GPUs present on a host, turn the prefix into a suffix:
>   `GPU_0__Nvidia RTX 3080` -> `Nvidia RTX 3080 (GPU 0)`
>   `GPU_1__Nvidia RTX 3070Ti` -> `Nvidia RTX 3070Ti (GPU 1)`
> 
> Advantages:
> 
> * no weird numbers when a host has only a single GPU, `friendly_name` is friendly
> * entity IDs always start with the make and model of the GPU
> * GPU index is present in entity ID if there are multiple GPUs
> * Multiple identical GPUs can be told apart
> 
> Disadvantages:
> 
> * A GPU's entity ID changes if another GPU is added to a host
> * ?


An easy solve to your disadvantage - always place the "GPU X" suffix at the end, regardless of how many GPUs there are. 

Seems there is nothing else outstanding with this PR, no?

fhoekstra commented:
That's a great solution!
I'll implement it as soon as I can make the time (been busy with a move).

wittypluck commented on homeassistant/components/glances/sensor.py:
```diff
@@ -234,6 +235,39 @@ class GlancesSensorEntityDescription(
+        key="mem",
+        type="gpu",
+        translation_key="gpu_memory_usage",
+        icon="mdi:memory",
```
Hello, I think you should move the icon to file `icons.json`

wittypluck commented on homeassistant/components/glances/sensor.py:
```diff
@@ -234,6 +235,39 @@ class GlancesSensorEntityDescription(
+        type="gpu",
+        translation_key="gpu_processor_usage",
+        native_unit_of_measurement=PERCENTAGE,
+        icon="mdi:expansion-card",
```
Same here

wittypluck commented on homeassistant/components/glances/sensor.py:
```diff
@@ -234,6 +235,39 @@ class GlancesSensorEntityDescription(
+        key="temperature",
+        type="gpu",
+        translation_key="temperature",
+        icon="mdi:thermometer",
```
Same here

wittypluck commented on homeassistant/components/glances/sensor.py:
```diff
@@ -234,6 +235,39 @@ class GlancesSensorEntityDescription(
+        type="gpu",
+        translation_key="fan_speed",
+        native_unit_of_measurement=PERCENTAGE,
+        icon="mdi:fan",
```
Same here

wittypluck commented on homeassistant/components/glances/sensor.py:
```diff
@@ -245,20 +279,13 @@ async def async_setup_entry(
-                            )
-                        )
+            _add_custom_sensors(coordinator, entities, sensor_type, sensors)
+        elif sensor_type == "gpu":
```
Wouldn't it be simpler to change the key formatting for gpu in python-glances-api and avoid having a special case for Gpu here ?
https://github.com/home-assistant-ecosystem/python-glances-api/blob/a73164393868d6c76421208953e771b9297209ed/glances_api/__init__.py#L185

wittypluck replied:
By the way I can submit the PR to python-glances-api if you want to got this way. I already have another PR open for `network` which will require a new version of the library.

fhoekstra replied:
Thanks, yes please! I was being impatient and hacky, hoping to get the change in before I would have less time. Wrong mindset. It is obviously better to change the formatting in the package itself than add complexity here.
Thank you for your offer to add the PR in there, I am still very busy.

wittypluck replied:
Hi @fhoekstra , I have drafted a pull request here:
https://github.com/home-assistant-ecosystem/python-glances-api/pull/35

The proposed format for the GPU name is:
`f"{sensor['name']} (GPU {sensor['gpu_id']})"`
which translates to : 
```
    "gpu": {
        "NVIDIA GeForce RTX 4080 (GPU 0)": {
            "mem": 13.333489176233513,
            "proc": 12,
            "temperature": 38,
            "fan_speed": 30,
        },
        "NVIDIA GeForce RTX 3080 (GPU 1)": {
            "mem": 8.41064453125,
            "proc": 26,
            "temperature": 51,
            "fan_speed": 0,
        },
    },
```
If you can have a quick look, I'll submit it to the library maintainer.

wittypluck replied:
PR submitted to python-glances-api

wittypluck replied:
Version 6.0 is now available with the formatting change on GPU name

wittypluck replied:
Version 0.6.0 has been merged into dev branch

fhoekstra replied:
Thanks for the quick action, I've rebased the branch. Also moving the icon definitions but have some trouble with my local dev setup. If I can't figure it out quickly, I could give you write access to the branch so I'm not blocking you any longer

wittypluck reviewed (commented):
Hello, a few comments as I just went through a similar PR review for Glances uptime sensor

fhoekstra committed:
PR feedback: move icons to icons.json

thanks to @wittypluck

fhoekstra committed:
Update glances snapshot with added GPU

fhoekstra commented:
Thanks for your help @wittypluck !
@joostlek I am sorry for tagging you and for not updating the status of this PR appropriately in the last months but now it seems ready to merge.
All feedback addressed, working locally, tests pass.
![image](https://github.com/home-assistant/core/assets/32362869/430c2218-675c-4ae2-b86c-aa54e448481d)
![image](https://github.com/home-assistant/core/assets/32362869/6b1aa437-5930-4e15-9b16-39515f96ecc6)

joostlek commented on homeassistant/components/glances/icons.json:
```diff
@@ -54,6 +54,15 @@
       "uptime": {
         "default": "mdi:clock-time-eight-outline"
+      },
+      "temperature": {
+        "default": "mdi:thermometer"
+      },
```
```suggestion
```
Is the default

joostlek reviewed (changes requested):
It looks good to go, but there I have this little comment (I saw it on your screenshot, might as well limit the precision directly).

Please delete all the snapshots and generate them again, Syrupy doesn't delete them when they aren't needed anymore, and I have a feeling that it has some old snapshots. After that, we can merge this

fhoekstra committed:
Remove JSON trailing comma

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

fhoekstra committed:
Limit precision of Glance GPU mem usage sensor

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

fhoekstra committed:
Clean up outdated snapshots

fhoekstra commented:
Thanks @joostlek for spotting those issues. I removed and regenerated the snapshots and it turned out to be necessary.
